src/transform/src/predicate_propagation.rs

```diff
@@ -0,0 +1,837 @@
+            p.is_literal_false()
+        } else {
+            false
+        }
+    }
+
+    fn is_literal_null(&self) -> bool {
+        if let Self::Predicate(p) = self {
+            p.is_literal_null()
+        } else {
+            false
+        }
+    }
+
+    /// Simplifies all the expressions within the Constraint.
+    fn reduce(&mut self, relation_type: &repr::RelationType) {
+        for expr in self.iter_expr_mut() {
+            expr.reduce(relation_type);
+        }
+    }
+
+    /// Maps all the expressions within the Constraint.
+    fn map_to_global(mut self, input_mapper: &JoinInputMapper, index: usize) -> Self {
+        for expr in self.iter_expr_mut() {
+            *expr = input_mapper.map_expr_to_global(expr.to_owned(), index);
+        }
+        self
+    }
+
+    /// Attempts to perform substitutions via `map` in all the expressions contained
+    /// in the constraint and returns `false` if the constraint cannot be lifted.
+    fn substitute(&mut self, map: &HashMap<MirScalarExpr, MirScalarExpr>) -> bool {
+        match self {
+            Self::Equivalence(class) => {
+                // Keep the class members that can find substitutions
+                *class = class
+                    .drain(..)
+                    .flat_map(|mut expr| {
+                        if substitute(&mut expr, map) {
+                            Some(expr)
+                        } else {
+                            None
+                        }
+                    })
+                    .collect::<Vec<_>>();
+                class.len() > 1
+            }
+            Self::Predicate(p) => substitute(p, map),
+        }
+    }
+
+    /// Optimizes all the expressions in the constraint by performing replacements.
+    /// Returns true if any replacement took place.
+    fn optimize(&mut self, predicates: &PredicateStructure) -> bool {
+        self.iter_expr_mut()
+            .map(|e| optimize(e, predicates))
+            .fold(false, |acc, v| acc || v)
+    }
+}
+
+impl std::fmt::Display for Constraint {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+        match self {
+            Self::Equivalence(class) => {
+                write!(f, "Equivalence[")?;
+                for (index, expr) in class.iter().enumerate() {
+                    write!(f, "{}{}", if index > 0 { " " } else { "" }, expr)?;
+                }
+                write!(f, "]")
+            }
+            Self::Predicate(p) => p.fmt(f),
+        }
+    }
+}
+
+/// Adds a constraint, expressed as a predicate, to the collection.
+///
+/// Equality predicates are converted into an equivalence and an IS NOT NULL predicate.
+fn append_constraint(constraints: &mut Vec<Constraint>, predicate: MirScalarExpr) {
+    if let MirScalarExpr::CallBinary {
+        expr1,
+        expr2,
+        func: BinaryFunc::Eq,
+    } = predicate
+    {
+        constraints.push(Constraint::Predicate(
```

Comment:
why are we asserting that `{expr1} is not null` here?

Reply:
`BinaryFunc::Eq` is a null rejecting expression, but `Constraint::Equivalence` models non-null-rejecting equality, so:

```
(binary_func_eq a b) ~= (equivalence a b) && (a is not null)
```